### PR TITLE
fix: patch removeEventListener to properly remove patched callbacks

### PR DESCRIPTION
**Pull request by johnbley**
We need to patch removeEventListener in order to compensate for wrapping callbacks passed to addEventListener.

### CODE REVIEW

File: plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts [hunk 1]
```diff
@@ -193,11 +194,36 @@ export class UserInteractionPlugin extends BasePlugin<unknown> {
...
+  private _patchRemoveEventListener() {
+    const plugin = this;
+    return (original: Function) => {
+      return function removeEventListenerPatched(
+        this: HTMLElement,
+        type: any,
+        listener: any,
+        useCapture: any
+      ) {
+        const wrappedListener = plugin._wrappedListeners.get(listener);
```
Comment (dyladan) [inline plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:217]: I would also remove the wrapped listener from the weakmap here

Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:217]: Good call; patching now.

**Comment by dyladan** [comment]
@rezakrimi the lint action is failing completely. Can you please look into this?

**Comment by rezakrimi** [comment]
> @rezakrimi the lint action is failing completely. Can you please look into this?

Sure.

**Comment by obecny** [comment]
Can you please provide more details what is the exact problem here ?, thx

**Comment by johnbley** [comment]
> Can you please provide more details what is the exact problem here ?, thx

Sure thing!  The added test shows that the addEventListener instrumentation actually breaks apps because it doesn't have the compensating logic for removeEventListener.  A patched callback is actually added as the event listener, and when the app tries to remove the original one, it's not found/ignored, leaving our patched callback present (along with unexpected calls to the underlying original one).  For some apps, this doesn't matter, but for others this can be arbitrarily bad (breaking functionality, thrown errors, infinite loops, etc. etc.).

**Comment by obecny** [comment]
it looks fine, the failing build might be connected with yesterday eslint merge, we have to check it @dyladan https://circleci.com/gh/open-telemetry/opentelemetry-js-contrib/1546

**Comment by dyladan** [comment]
> it looks fine, the failing build might be connected with yesterday eslint merge, we have to check it @dyladan [circleci.com/gh/open-telemetry/opentelemetry-js-contrib/1546](https://circleci.com/gh/open-telemetry/opentelemetry-js-contrib/1546)
> EDIT:
> I might be wrong with the link, but when checking the lint in this PR something else is failing not related to this PR

It is definitely related. eslint isn't running at all because of a gts dependency error.

**Comment by rezakrimi** [comment]
> > it looks fine, the failing build might be connected with yesterday eslint merge, we have to check it @dyladan [circleci.com/gh/open-telemetry/opentelemetry-js-contrib/1546](https://circleci.com/gh/open-telemetry/opentelemetry-js-contrib/1546)
> > EDIT:
> > I might be wrong with the link, but when checking the lint in this PR something else is failing not related to this PR
> 
> It is definitely related. eslint isn't running at all because of a gts dependency error.

yes the caching was wrong, I'll make a PR today to fix it

File: plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts [hunk 2]
```diff
@@ -193,11 +194,37 @@ export class UserInteractionPlugin extends BasePlugin<unknown> {
             return listener.apply(target, args);
           }
         };
+        plugin._wrappedListeners.set(listener, patchedListener);
```
Comment (lykkin) [inline plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:197]: It looks like the listeners are assumed to be unique. If so this wouldn't handle reusing the same listener, though I don't know if that's an edge case we want to worry about.

**Comment by rezakrimi** [comment]
#159 should fix the lint check.

**Comment by dyladan** [comment]
@johnbley if you update from master now the lint should run properly

File: plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts (hunk 2 above)
Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:197]: No, this is an important point, thanks.  In particular I'm going to tweak it to properly support
```
  element.addEventListener('type1', listener);
  element.addEventListener('type2', listener);
  element.removeEventListener('type1', listener);
```
Which the current instrumentation does not properly support.  Thanks for the careful review!

**Comment by johnbley** [comment]
> @johnbley if you update from master now the lint should run properly

Will do, but please hold on merging until I find the time to properly handle some edge cases better.

File: plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts (hunk 2 above)
Comment (obecny) [reply plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:197]: When I start thinking more about this it is becoming more complicated when I originally thought and it will require to make few other changes because when you call AddEventListener with the same callback and type it will overwrite previous callback. Also it should only create span for click event. So when saving patchedListener you should also save the type

If I haven't missed anything the patch function should look like
```
  private _wrappedListeners = new WeakMap<Function, Record<string, ()=>{}>>();
  private _patchElement() {
    const plugin = this;
    return (original: Function) => {
      return function addEventListenerPatched(
        this: HTMLElement,
        type: string,
        listener: any,
        useCapture: any
      ) {
        const patchedListener = (...args: any[]) => {
          const target = this;
          if (type.toLowerCase() !== 'click') {
            return listener.apply(target, args);
          }
          const span = plugin._createSpan(target, 'click');
          if (span) {
            return plugin._tracer.withSpan(span, () => {
              const result = listener.apply(target, args);
              // no zone so end span immediately
              span.end();
              return result;
            });
          } else {
            return listener.apply(target, args);
          }
        };
        let listeners = plugin._wrappedListeners.get(original);
        if (!listeners) {
          listeners = {};
          listeners[type] = patchedListener;
          plugin._wrappedListeners.set(original, listeners);
        } else if (listeners[type]) {
          this.removeEventListener(type, listeners[type]);
          delete listeners[type];
        }
        listeners[type] = patchedListener;
        // not sure if the below line is yet needed
        plugin._wrappedListeners.set(original, listeners);
        return original.call(this, type, patchedListener, useCapture);
      };
    };
  }
```

After this you will also need to modify the patch for `removeEventListener` so that it will take into consideration the type.

Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:197]: This is very similar to where I landed, with the added bit that, of course, the element it's on is a part of the registration.  I've got a separate patch for better handling of the event type and would rather not conflate the two.

**Comment by johnbley** [comment]
Looks like the lint step in the build is failing again, but this time with out-of-date code... I ran lint:fix and pushed those changes.

**Comment by rezakrimi** [comment]
> Looks like the lint step in the build is failing again, but this time with out-of-date code... I ran lint:fix and pushed those changes.

That's strange. Same thing happened to my PR on the main opentelemetry-js repo. For me it just got resolved after pushing another commit.

File: plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts [hunk 3]
```diff
@@ -89,6 +89,40 @@ describe('UserInteractionPlugin', () => {
       context.disable();
     });
 
+    it('should not break removeEventListener', () => {
+      let called = false;
+      const listener = function () {
+        called = true;
+      };
+      // add same listener three different ways
```
Comment (obecny) [inline plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: I'm curious about one scenario, to check that can you modify this a bit to test the case with overwriting the previous event with different callback
```
let called = false;
let called2 = false;
const listener = function () {
  called = true;
};
const listener2 = function () {
  called2 = true;
};
document.body.addEventListener('bodyEvent', listener);
document.body.addEventListener('bodyEvent', listener2);
document.body.addEventListener('bodyEvent2', listener);

// now fun
document.body.dispatchEvent(new Event('bodyEvent'));
assert.strictEqual(called, false);
assert.strictEqual(called2, true);
```

Comment (obecny) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: fixed my example

Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: Yep, good catch and an easy fix for this one based on the tracking state.  Pushing fix momentarily.

Comment (obecny) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: There is one more thing a bonus :). Currently you can call `addEventListener` providing a 3rd param `options`. In options you can set `once:true` which means the event will be removed automatically. It might be worth to test that too.

Comment (obecny) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]:
```
let called2 = 0;
const listener2 = function () {
  called2++;
};
document.body.addEventListener('bodyEvent', listener2);
document.body.addEventListener('bodyEvent', listener2, { once: true });

// now fun
document.body.dispatchEvent(new Event('bodyEvent'));
assert.strictEqual(called2, 1);
document.body.dispatchEvent(new Event('bodyEvent'));
assert.strictEqual(called2, 2); // error ?
```

Comment (obecny) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: and you can revert it add `once` first and then add without options, heh small things and many edge cases

Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: Argh.  There's also the case that once:true shouldn't prevent us from allowing it to be adding later (it's not really a double-register).  Will dive in and fix this one tomorrow.  Thanks!

Comment (obecny) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: thx for taking this into deep details :)

Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:97]: Couldn't leave it alone; handled the `once:true` callbacks properly being removed when an event is dispatched.

File: plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts [hunk 4]
```diff
@@ -89,6 +89,73 @@ describe('UserInteractionPlugin', () => {
...
+      let callCount = 0;
+      const listener = function () {
+        callCount++;
+      };
+      // addEventListener semantics treat the second call as a no-op
+      document.body.addEventListener('bodyEvent', listener, { once: true });
+      document.body.addEventListener('bodyEvent', listener); // considered a double-register
+      document.body.dispatchEvent(new Event('bodyEvent'));
+      assert.strictEqual(callCount, 1);
+      // now that it's been dispatched once, it's been removed; should be able to re-add
```
Comment (obecny) [inline plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:153]: Nit: I would add here extra call and check
```
document.body.dispatchEvent(new Event('bodyEvent'));
assert.strictEqual(callCount, 1);
```

File: plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts [hunk 5]
```diff
@@ -89,6 +89,73 @@ describe('UserInteractionPlugin', () => {
...
+      document.body.addEventListener('bodyEvent', listener);
+      document.body.dispatchEvent(new Event('bodyEvent'));
+      assert.strictEqual(callCount, 2);
```
Comment (obecny) [inline plugins/web/opentelemetry-plugin-user-interaction/test/userInteraction.nozone.test.ts:156]: Nit: and the same here

```
document.body.dispatchEvent(new Event('bodyEvent'));
assert.strictEqual(callCount, 3);
```

File: plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts [hunk 6]
```diff
@@ -165,6 +170,61 @@ export class UserInteractionPlugin extends BasePlugin<unknown> {
...
+   */
+  private addPatchedListener(
+    on: HTMLElement,
+    type: string,
+    listener: Function,
+    wrappedListener: Function
+  ): boolean {
+    let listener2Type = this._wrappedListeners.get(listener);
+    if (!listener2Type) {
+      listener2Type = new Map();
```
Comment (lykkin) [inline plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:184]: Is this going to be coerced into a `WeakMap` when you set it on `_wrappedListeners`?

Comment (johnbley) [reply plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:184]: No, though making the HTMLElement reference weak is a good idea.  Unfortunately I don't see a way to do it since there are no iterators/size/isEmpty methods available on WeakMap and without it we'd *definitely* leak.  I'll ponder this but I'd like to get this fix in as it is since it's clearly better than the current breaking of functional behavior.

**Comment by johnbley** [comment]
> great job, thx for investigating this further :)

Thanks to all for spotting further issues and edge cases!

File: plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts (hunk 6 above)
Comment (lykkin) [reply plugins/web/opentelemetry-plugin-user-interaction/src/userInteraction.ts:184]: Awesome, thanks for the clarification!
